File: server/src/routes/events.js
import express from 'express';
import { nanoid } from 'nanoid';
import db from '../db/database.js';
import { requireAuth } from '../middleware/auth.js';

const router = express.Router();
router.use(requireAuth);

router.get('/', (req, res) => {
  const { from, to } = req.query;
  let sql = 'SELECT * FROM events WHERE user_id = ?';
  const params = [req.user.id];
  if (from) { sql += ' AND start_at >= ?'; params.push(from); }
  if (to) { sql += ' AND start_at <= ?'; params.push(to); }
  sql += ' ORDER BY start_at ASC';
  const events = db.prepare(sql).all(...params);
  res.json({ events });
});

router.post('/', (req, res) => {
  const { title, description, start_at, end_at, location, color } = req.body || {};
  if (!title || !start_at || !end_at) {
    return res.status(400).json({ error: 'title, start_at, end_at are required' });
  }
  const id = nanoid();
  db.prepare(
    'INSERT INTO events (id, user_id, title, description, start_at, end_at, location, color, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'
  ).run(id, req.user.id, title, description || null, start_at, end_at, location || null, color || 'indigo', Date.now());
  const event = db.prepare('SELECT * FROM events WHERE id = ?').get(id);
  res.json({ event });
});

router.patch('/:id', (req, res) => {
  const existing = db.prepare('SELECT * FROM events WHERE id = ? AND user_id = ?').get(req.params.id, req.user.id);
  if (!existing) return res.status(404).json({ error: 'Event not found' });

  const fields = ['title', 'description', 'start_at', 'end_at', 'location', 'color'];
  const updates = [];
  const values = [];
  for (const f of fields) {
    if (f in req.body) { updates.push(`${f} = ?`); values.push(req.body[f]); }
  }
  if (!updates.length) return res.json({ event: existing });
  values.push(req.params.id, req.user.id);
  db.prepare(`UPDATE events SET ${updates.join(', ')} WHERE id = ? AND user_id = ?`).run(...values);
  const event = db.prepare('SELECT * FROM events WHERE id = ?').get(req.params.id);
  res.json({ event });
});

router.delete('/:id', (req, res) => {
  db.prepare('DELETE FROM events WHERE id = ? AND user_id = ?').run(req.params.id, req.user.id);
  res.json({ ok: true });
});

export default router;

File: server-py/app/api/events.py
import time

from fastapi import APIRouter, Depends, HTTPException, Request
from nanoid import generate as nanoid

from app.db import execute, query_all, query_one, row_to_dict
from app.deps import require_auth
from app.models.event import EventCreate

router = APIRouter(prefix="/events")


def _now_ms() -> int:
    return int(time.time() * 1000)


@router.get("")
def list_events(request: Request, user=Depends(require_auth)):
    # `from` is a Python keyword, so read query params directly.
    from_value = request.query_params.get("from")
    to_value = request.query_params.get("to")
    sql = "SELECT * FROM events WHERE user_id = ?"
    params: list = [user["id"]]
    if from_value:
        sql += " AND start_at >= ?"; params.append(from_value)
    if to_value:
        sql += " AND start_at <= ?"; params.append(to_value)
    sql += " ORDER BY start_at ASC"
    return {"events": [row_to_dict(r) for r in query_all(sql, tuple(params))]}


@router.post("")
def create_event(body: EventCreate, user=Depends(require_auth)):
    if not body.title or not body.start_at or not body.end_at:
        raise HTTPException(400, "title, start_at, end_at are required")
    event_id = nanoid()
    execute(
        """INSERT INTO events (id, user_id, title, description, start_at, end_at, location, color, created_at)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (event_id, user["id"], body.title, body.description, body.start_at, body.end_at,
         body.location, body.color or "indigo", _now_ms()),
    )
    return {"event": row_to_dict(query_one("SELECT * FROM events WHERE id = ?", (event_id,)))}


_PATCH_FIELDS = ("title", "description", "start_at", "end_at", "location", "color")


@router.patch("/{event_id}")
def update_event(event_id: str, body: dict, user=Depends(require_auth)):
    existing = query_one("SELECT * FROM events WHERE id = ? AND user_id = ?", (event_id, user["id"]))
    if not existing:
        raise HTTPException(404, "Event not found")
    updates: list[str] = []
    values: list = []
    for f in _PATCH_FIELDS:
        if f in body:
            updates.append(f"{f} = ?"); values.append(body[f])
    if not updates:
        return {"event": row_to_dict(existing)}
    values.extend([event_id, user["id"]])
    execute(
        f"UPDATE events SET {', '.join(updates)} WHERE id = ? AND user_id = ?",
        tuple(values),
    )
    return {"event": row_to_dict(query_one("SELECT * FROM events WHERE id = ?", (event_id,)))}


@router.delete("/{event_id}")
def delete_event(event_id: str, user=Depends(require_auth)):
    execute("DELETE FROM events WHERE id = ? AND user_id = ?", (event_id, user["id"]))
    return {"ok": True}
